idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/references/FirReferenceResolveHelper.kt

```diff
@@ -374,42 +375,104 @@ internal object FirReferenceResolveHelper {
+            var classId =
+                if ((referencedClass as? FirRegularClass)?.isCompanion == true &&
+                    (qualifiedAccess.selectorExpression as? KtNameReferenceExpression)?.getReferencedName() != "Companion"
+                ) {
+                    // Remove the last "Companion" part if the qualified access does not contain it. This is needed because the "Companion"
+                    // part is optional.
+                    referencedClass.classId.outerClassId ?: return referencedSymbolsByFir
+                } else {
+                    referencedClass.classId
+                }
+            val qualifiedAccessSegments = qualifiedAccess.fqNameSegments() ?: return referencedSymbolsByFir
+            assert(classId.asSingleFqName().pathSegments().takeLast(qualifiedAccessSegments.size)
+                       .map { it.identifierOrNullIfSpecial } == qualifiedAccessSegments) {
+                "Referenced classId $classId should end with qualifiedAccess expression ${qualifiedAccess.text} "
+            }
 
-        var parent = expression.parent as? KtDotQualifiedExpression
-        // Distinguish A.foo() from A(.Companion).foo()
-        // Make expression.parent as? KtDotQualifiedExpression local function
-        while (parent != null) {
-            val selectorExpression = parent.selectorExpression ?: break
-            if (selectorExpression === expression) {
-                parent = parent.parent as? KtDotQualifiedExpression
-                continue
+            // Handle nested classes.
+            while (true) {
+                if (expression === qualifiedAccess.selectorExpression) {
+                    return listOfNotNull(classId.toTargetPsi(session, symbolBuilder))
+                }
+                val outerClassId = classId.outerClassId
+                val receiverExpression = qualifiedAccess.receiverExpression
+                if (receiverExpression !is KtDotQualifiedExpression) {
+                    // If the receiver is not a KtDotQualifiedExpression, it means we are hitting the end of nested receivers.
+                    if (receiverExpression == expression) {
+                        // If there is still an outer class, then return symbol of that class
+                        outerClassId?.let { return listOfNotNull(it.toTargetPsi(session, symbolBuilder)) }
+                        // Otherwise, it should be a package, so we return that
+                        return listOfNotNull(symbolBuilder.createPackageSymbolIfOneExists(classId.packageFqName))
+                    } else {
+                        // This is unexpected. The code probably contains some weird structures. In this case, we just fail the resolution
+                        // with zero results.
+                        return emptyList()
+                    }
+                }
+                qualifiedAccess = receiverExpression
+                if (outerClassId == null) break
```

Comment:
Nit: it might be a personal preference, but I wonder if it's better to avoid `while(true) { ... break }` ? :) How about changing the loop condition to `classId != null` instead?

Reply:
Oh wait... you need `classId` to be the outer-most class and then go to the next loop to handle package names... It'd be good to have some comments before/after this first loop to explain the invariant of `classId` then (in addition to "in-sync" state).

Reply:
OK, let me refactor this code to remove `while(true)`.